libyul/backends/evm/EVMCodeTransform.cpp

```diff
@@ -182,16 +182,24 @@ void CodeTransform::operator()(VariableDeclaration const& _varDecl)
 			else
 				m_variablesScheduledForDeletion.insert(&var);
 		}
-		else if (m_unusedStackSlots.empty())
-			atTopOfStack = false;
 		else
 		{
-			auto slot = static_cast<size_t>(*m_unusedStackSlots.begin());
-			m_unusedStackSlots.erase(m_unusedStackSlots.begin());
-			m_context->variableStackHeights[&var] = slot;
-			if (size_t heightDiff = variableHeightDiff(var, varName, true))
-				m_assembly.appendInstruction(evmasm::swapInstruction(static_cast<unsigned>(heightDiff - 1)));
-			m_assembly.appendInstruction(evmasm::Instruction::POP);
+			bool foundUnusedSlot = false;
+			for (auto it = m_unusedStackSlots.begin(); it != m_unusedStackSlots.end(); ++it)
+			{
+				if (m_assembly.stackHeight() - *it > 17)
+					continue;
+				foundUnusedSlot = true;
+				auto slot = static_cast<size_t>(*it);
+				m_unusedStackSlots.erase(it);
+				m_context->variableStackHeights[&var] = slot;
+				if (size_t heightDiff = variableHeightDiff(var, varName, true))
+					m_assembly.appendInstruction(evmasm::swapInstruction(static_cast<unsigned>(heightDiff - 1)));
+				m_assembly.appendInstruction(evmasm::Instruction::POP);
```

Comment:
Can you add a test for this change?

Reply:
Ah well :-). That's a bit annoying to do, but I can try :-). I should have kept the fuzzer examples that triggered this and added them as tests in the first place :-).
It's clear what it does and why it's required, though, right? It tried to reuse slots that were immediately unreachable, so swapping into it immediately failed.
The main issue is that this change here alone cannot prevent this kind of thing from happening in all cases - i.e. reusing any particular slot here might non-locally have this variable or some other variable (that will no longer be able to choose this slot) be unreachable down the road and it's impossible to determine that here - and I don't see any way to solve this problem - that's actually the main reason for writing a new code transform :-).

Reply:
I added a test case that will trigger a "stack too deep" without this change, but will not with this change. But solely for the code transform without the stack limit evader.
Is that enough :-)?
Ideally, I would have extracted this change and sold it as a change only to the code transform itself :-).

For that particular test, I think just accepting the "stack too deep" and moving the variable to memory might still work - building a test case, for which that also fails without this change, but doesn't with it, is a bit harder. If you want, I can spend some time trying, though, let me know...

Reply:
Ok thanks. Mostly wanted to see where this is useful.